File: flutter_project/lib/models/clock_settings.dart
import 'package:flutter/material.dart';

/// Model representing all customizable clock preferences.
///
/// Educational Note:
/// - In Flutter/Dart, immutable data classes with copyWith() are the gold standard.
/// - When an attribute changes, we emit a new ClockSettings instance via copyWith(),
///   which triggers reactive widget rebuilds cleanly without mutation side-effects.
class ClockSettings {
  final bool is24Hour;
  final bool showSeconds;
  final bool showDate;
  final String fontStyle;
  final Color clockColor;
  final String bgType; // 'solid', 'gradient', or 'image'
  final Color bgSolidColor;
  final int bgGradientIndex;
  final String? customImagePath;
  final String clockSize; // 'small', 'medium', 'large', 'xlarge'
  final String clockPosition; // 'center', 'top', 'bottom'
  final String customNote;
  final bool showNote;

  const ClockSettings({
    this.is24Hour = false,
    this.showSeconds = true,
    this.showDate = true,
    this.fontStyle = 'Digital',
    this.clockColor = const Color(0xFF00E5FF),
    this.bgType = 'solid',
    this.bgSolidColor = const Color(0xFF0D1117),
    this.bgGradientIndex = 0,
    this.customImagePath,
    this.clockSize = 'large',
    this.clockPosition = 'center',
    this.customNote = 'Stay Focused 🎯',
    this.showNote = true,
  });

  static const List<String> allFonts = [
    'Digital',
    'Minimal',
    'Thin',
    'Bold',
    'Futuristic',
    'Seven Segment',
    'Rounded',
    'Retro',
    'Monospace',
    'Serif',
  ];

  ClockSettings copyWith({
    bool? is24Hour,
    bool? showSeconds,
    bool? showDate,
    String? fontStyle,
    Color? clockColor,
    String? bgType,
    Color? bgSolidColor,
    int? bgGradientIndex,
    String? customImagePath,
    String? clockSize,
    String? clockPosition,
    String? customNote,
    bool? showNote,
  }) {
    return ClockSettings(
      is24Hour: is24Hour ?? this.is24Hour,
      showSeconds: showSeconds ?? this.showSeconds,
      showDate: showDate ?? this.showDate,
      fontStyle: fontStyle ?? this.fontStyle,
      clockColor: clockColor ?? this.clockColor,
      bgType: bgType ?? this.bgType,
      bgSolidColor: bgSolidColor ?? this.bgSolidColor,
      bgGradientIndex: bgGradientIndex ?? this.bgGradientIndex,
      customImagePath: customImagePath ?? this.customImagePath,
      clockSize: clockSize ?? this.clockSize,
      clockPosition: clockPosition ?? this.clockPosition,
      customNote: customNote ?? this.customNote,
      showNote: showNote ?? this.showNote,
    );
  }
}

File: app/src/main/java/com/example/data/ClockSettings.kt
package com.example.data

import androidx.room.Entity
import androidx.room.PrimaryKey

/**
 * ClockSettings
 *
 * This entity model represents the user's saved clock preferences stored locally in the Room database.
 * Each property has beginner-friendly default values.
 *
 * Educational Note:
 * - @Entity tells Room (SQLite abstraction) to create a database table named "clock_settings".
 * - @PrimaryKey(autoGenerate = false) ensures we only store a single configuration row (id = 1).
 *   Whenever settings change, we update this exact row.
 */
@Entity(tableName = "clock_settings")
data class ClockSettings(
    @PrimaryKey
    val id: Int = 1,

    // Time Formatting
    val is24Hour: Boolean = false,      // false = 12-hour format with AM/PM, true = 24-hour format
    val showSeconds: Boolean = true,    // Toggle displaying :SS
    val showDate: Boolean = true,       // Toggle displaying the date line

    // Custom Note
    val customNote: String = "",        // Custom message or note displayed with clock
    val showNote: Boolean = true,       // Toggle displaying the custom note

    // Visual Customization
    val fontStyle: String = FONT_DIGITAL, // 1 of 10 supported font styles
    val colorHex: Long = 0xFF00FFCC,       // Neon Cyan default (ARGB as Long)

    // Background Customization
    val bgType: String = BG_SOLID,      // "solid", "gradient", or "image"
    val bgSolidHex: Long = 0xFF0D1117,  // Dark cyber canvas default
    val bgGradientIndex: Int = 0,       // Preset gradient index (0 to 6)
    val customImageUri: String? = null, // URI string if user selected a gallery picture

    // Layout
    val clockSize: String = SIZE_LARGE,       // "small", "medium", "large", "xlarge"
    val clockPosition: String = POS_CENTER    // "center", "top", "bottom"
) {
    companion object {
        // 10 Clock Font Styles
        const val FONT_DIGITAL = "Digital"
        const val FONT_MINIMAL = "Minimal"
        const val FONT_THIN = "Thin"
        const val FONT_BOLD = "Bold"
        const val FONT_FUTURISTIC = "Futuristic"
        const val FONT_SEVEN_SEGMENT = "Seven Segment"
        const val FONT_ROUNDED = "Rounded"
        const val FONT_RETRO = "Retro"
        const val FONT_MONOSPACE = "Monospace"
        const val FONT_SERIF = "Serif"

        val ALL_FONTS = listOf(
            FONT_DIGITAL,
            FONT_MINIMAL,
            FONT_THIN,
            FONT_BOLD,
            FONT_FUTURISTIC,
            FONT_SEVEN_SEGMENT,
            FONT_ROUNDED,
            FONT_RETRO,
            FONT_MONOSPACE,
            FONT_SERIF
        )

        // Background Types
        const val BG_SOLID = "solid"
        const val BG_GRADIENT = "gradient"
        const val BG_IMAGE = "image"

        // Clock Sizes
        const val SIZE_SMALL = "small"
        const val SIZE_MEDIUM = "medium"
        const val SIZE_LARGE = "large"
        const val SIZE_XLARGE = "xlarge"

        // Clock Positions
        const val POS_TOP = "top"
        const val POS_CENTER = "center"
        const val POS_BOTTOM = "bottom"

        // Default instance
        val DEFAULT = ClockSettings()
    }
}
